Add Reusable block save button, snackbar on save and Welcome Guide

thisissandip opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/32020

## Description
Added a Save Button for Reusable Blocks in Blocks Toolbar.  If there are no changes, the save button is disabled. When a change is made in that particular reusable block. One can save the changes from the block toolbar. Once saved, the user is notified with a snackbar about the save. The snackbar's 'Learn More' opens the Reusable Block Welcome Guide which also has a link to external documentation.

Since the image (gif) for the welcome guide is not ready yet. I have kept a plain background in its place.
I will add the image whenever it's made available.

## How has this been tested?
1. Select a Reusable block
2. Make Changes
3. Save the Reusable Block from Toolbar
4. A snackbar should appear at the bottom left 
5. Click on Snackbar's Learn More. This should open the Welcome guide.


## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/69596988/120833132-117f3580-c57f-11eb-8a39-1626f89f28e5.mov




## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

thisissandip committed:
Add Reusable block save button, snackbar on save and Welcome Guide

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -84,6 +97,56 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+
+	// Get all the unsaved Records
+	const dirtyRecords = useSelect( ( select ) => {
+		return select( coreStore ).__experimentalGetDirtyEntityRecords();
```
Since we're checking edits (dirty records) for a single entity, I think we can use the `hasEditsForEntityRecord` selector here.

Something like this:

```js
select( coreStore ).hasEditsForEntityRecord( 'postType', 'wp_block', ref );
```

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -84,6 +97,56 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+	}, [] );
+
+	const selectedBlock = useSelect( ( select ) => {
+		return select( blockEditorStore ).getSelectedBlock();
```
This is used to get `ref` later in code, but we already have a `ref` for the current block.

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -84,6 +97,56 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+
+	// save reusable block
+	const saveDirtyRecords = ( selectedBlockref ) => {
+		// only save the selected block records
```
I think we can use the `saveEditedEntityRecord` action for this.

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -84,6 +97,56 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+	};
+
+	const showSnackbar = () => {
+		createNotice( 'success', 'Reusable block Saved.', {
```
Missing l10n.

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -84,6 +97,56 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+			isDismissible: true,
+			actions: [
+				{
+					label: 'Learn More',
```
Same here with missing l10n.

Mamaduka reviewed (changes requested):
Thanks for working on this PR, @thisissandip.

I left few comments that I think would be nice to address before merging.

I'm not 100% sure about the Welcome Guide. Maybe the "Learn More" in snackbar should link to external documentation?

cc @critterverse

critterverse commented:
Nice! Thanks for working on this @thisissandip.

I can create the GIF asset for the Welcome Guide — will follow up on this next week.

A couple of notes on the PR:

- There are some changes that don't trigger the "Save" button in the toolbar to activate — I was seeing this when changing between different Duotone colors on an inner Image block or turning on/off the "toggle full height" option, for example. It seems to trigger it _sometimes_ but not every time
- Can we adjust the text in the snackbar to remove the capital letters on the words "saved" and "more." It should look like: "Reusable block saved. Learn more"
- Let's change the CTA at the bottom of the Welcome Guide from "Get started" to "Got it"
- I think we can remove the link to the external support resource within the Guide for now until we have some good support documentation to link out to

thisissandip committed:
Update the way to check dirty records and Fix l10n

thisissandip committed:
Remove External Link and Update Finish Button Text in Welcome Gudie

thisissandip commented:
@Mamaduka Thank you for the review. I have made the required changes. Let me know if this looks good! 😄

thisissandip commented:
@critterverse Thank you for taking a look, I have updated the code. Those issues are now fixed. 
Also, I have made the required changes to the snackbar and welcome guide.

https://user-images.githubusercontent.com/69596988/121701608-b4d7c980-caee-11eb-8f57-465a77164ce7.mov


Looking forward to that GIF! 😄

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -84,6 +91,40 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+	// local states for gudie modal
+	const [ isGudieOpen, setIsGudieOpen ] = useState( false );
+
+	// check if block has unsaved records
+	const isBlockDirty = useSelect( ( select ) => {
+		return select( coreStore ).hasEditsForEntityRecord(
+			'postType',
+			'wp_block',
+			ref
+		);
+	}, [] );
```
Let's move this in top `useSelect`; it's selecting from the same store anyways.

P.S. I know the term "dirty" is used a lot in this context across the codebase, but I think we can use `hasEdits` or something similar.  This is totally up to you.

thisissandip replied:
Thank you for the suggestion! 
Even I think `hasEdits` is much better. I am changing the selector to `hasEdits` and renaming the save function to `saveEditedRecords`

Mamaduka replied:
Awesome.

Don't forget to move `hasEdits` selector into the top `useSelect`.

https://github.com/WordPress/gutenberg/blob/8c4282eae5882e87a518b361993ea3a8cbb90c51/packages/block-library/src/block/edit.js#L35-L55

thisissandip replied:
Added another commit! 😅  Thanks for the reminder!

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -126,6 +167,20 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
 					</ToolbarGroup>
+
+					<ToolbarButton
+						className="is-primary"
```
We can use the `isPrimary` prop for this.

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -126,6 +167,20 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+							margin: '7px 0px 7px 7px',
+							height: '34px',
+						} }
+						onClick={ () => saveDirtyRecords() }
```
You can just pass `saveDirtyRecords` here. No need for another function.

Mamaduka commented on packages/block-library/src/block/edit.js:
```diff
@@ -126,6 +167,20 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+						showTooltip
+						isDisabled={ ! isBlockDirty }
+					>
+						Save
```
It looks like we missed translation.

Mamaduka reviewed (dismissed):
Thanks for the follow-up, @thisissandip.

I left few minor comments, and I think we're good to land after this.

I'm still not 100% sure about the "Welcome Guide." In my experience, editors display similar guides when the user accesses the screen for the first time.

I general, I think it's a nice addition, just not sure if it's the right place. So I would love to get another opinion before we move forward with this.

cc @jasmussen

jasmussen commented:
Thanks for the PR, thanks for the ping. This is what I see:

![this](https://user-images.githubusercontent.com/1204802/122031869-64eb5200-cdcf-11eb-8597-7cb68ec69ece.gif)

I love the work @critterverse has done here, and this PR sticks to that very impressively, good work all around.

A few thoughts:

- The focus style for the save button is a white square inside. Seems like it should just be the same focus style as other primary buttons.
- The animated asset and contents of the "guide screen" is missing — almost certainly known, but just obviously something we need in place for this to land.
- The Save button is grouped with the ellipsis button. Which never occurred to me as an option — I've always thought of the ellipsis as always needing [its own group](https://make.wordpress.org/core/2021/04/23/core-editor-improvement-refining-the-block-toolbar/). However I can see the high level sense it makes: it can be thought of as part of the "More" group. I would probably put the Save button in its own "Other" group, but I also appreciate the desire to use borders and separators sparingly, so take it only as a thought!

It seems the primary open question is whether it's okay to open a reusable blocks guide from the snackbar, compared to opening it automatically when you edit a reusable block the first time. In principle @Mamaduka is right, for the time being we've only done the latter. At the same time, I appreciate @critterverse's solution on a personal level: whenever I see a "welcome guide" like this, I immediately and quickly dismiss, almost as if it was an advertisement. For me, the extra friction of a click is appreciated.

I can understand it also might be enough friction that no-one will ever see that guide and simply ignore that "Read more" link. But since this is working well, it seems fine to land this behavior and get a feel for it in the plugin: we can always double back and tweak it if it doesn't work as well in practice.

Thanks again for the work here! 👏

Mamaduka commented:
Thanks for the feedback, @jasmussen.

> whenever I see a "welcome guide" like this, I immediately and quickly dismiss, almost as if it was an advertisement.

I've also acquired this habit 😄  , but I assume this is helpful for new users who don't spend every day in the editor.

Welcome Modal assets are now hosted on WP.org. I'll create a new ticket on meta once we're close to land this PR.

paaljoachim commented:
Thanks for working on this feature, @thisissandip 

Various adjustments / new features to the Reusable block needs to also be considered for possible backporting to WP 5.7.3.
To fix various complaints that we have received in relation to Reusable blocks in WP 5.7.
@peterwilsoncc 

A question that comes up...
Should there also be a discard/undo change button (using text or an undo icon) next to the save button? As if a user makes a mistake it also needs to be just as easy to undo and not save the change. 

Btw I do think there should be a vertical black separator line between the Save button and the ellipsis (3 dot) drop down menu. As the ellipsis drop down should be separate from other toolbar icons.

------

Associated issue:
Reusable block: Add a lock to the parent toolbar to protect the inner content from accidental editing.
https://github.com/WordPress/gutenberg/issues/32461
Is this also an issue you are able to work on, @thisissandip ?

Another issue where I am rethinking various Reusable block features:
https://github.com/WordPress/gutenberg/issues/31163

thisissandip committed:
Move Save Button to it's own group

thisissandip committed:
move hasEdits to top useSelect

thisissandip commented:
I have moved the Save button to its own group as suggested by @jasmussen and @paaljoachim. Also, made the suggested changes by @Mamaduka. Thank you guys for reviewing it! 😄 

https://user-images.githubusercontent.com/69596988/122080201-b8838d00-ce1b-11eb-80bb-751c0c067511.mov

thisissandip commented:
@paaljoachim Thank you for taking a look 😄  

I just went through issue #32461. 
I have successfully implemented the unlock and lock feature. Should I create another PR for it or extend it in this PR?

paaljoachim commented:
Hi @thisissandip 

That is awesome!

Yes, do please create a new PR for the lock/unlock feature. 
We can then continue discussing the specific feature in the PR.

critterverse commented:
Thanks for the feedback, all! @jasmussen's note about adding the additional divider to the toolbar is a great call and the change you already made there looks perfect, @thisissandip. Just wanted to highlight one other note that I think we should take a look at:

>The focus style for the save button is a white square inside. Seems like it should just be the same focus style as other primary buttons.

The default focus style looks like this for reference (see also [this link](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=39%3A0) to the relevant section of WP Design Library):

![primary-button-style-focus](https://user-images.githubusercontent.com/77359364/122111931-48124700-cdee-11eb-8e4b-e6d0bd6217d2.jpg)

Here's what I came up with for the GIF asset (it only plays once so you might want to open in a new tab and refresh a few times):

![welcome-reusable](https://user-images.githubusercontent.com/77359364/122108877-bc4aeb80-cdea-11eb-8e8a-152a1b30d46f.gif)

It would be great to get some additional eyes on this — @pablohoneyhoney, does this asset look ok to you to use in a new Welcome Guide to introduce reusable blocks? This implementation is based on some of your previous notes from #29871. Please see the most recent video above in https://github.com/WordPress/gutenberg/pull/32464#issuecomment-861593761 for context!

>Welcome Modal assets are now hosted on WP.org. I'll create a new ticket on meta once we're close to land this PR.

Thanks @Mamaduka! I was just looking at [this Trac ticket](https://meta.trac.wordpress.org/ticket/5741) and tried to match the asset as closely as possible to what was uploaded there. Looks like there is also a SVG static asset needed which I can supply as well pending any feedback :)

Mamaduka commented:
Thanks, @critterverse. If you can provide assets similar to the ones in the Trac ticket, that would be perfect.

@thisissandip it looks like we have ESLint errors. We should fix those to pass all the required checks.

thisissandip committed:
Fix ESLint Errors

paaljoachim commented:
The lock/unlock PR will affect this PR, as one should not be able to save a change as long as the lock is closed/locked.
Which means the lock/unlock PR needs to be merged before this PR is merged.

Mamaduka commented:
>The lock/unlock PR will affect this PR, as one should not be able to save a change as long as the lock is closed/locked.
Which means the lock/unlock PR needs to be merged before this PR is merged.

Not necessarily. The "Save" button is disabled if there are no edits, and I think this will be the case when the block is "locked."

paaljoachim commented:
Good point George!

Let us get the lock/unlock PR in place first, so that we can also add some explorations in it before this PR merged. 
I also believe we need to get this PR in front of more eyes. I happen to come across it while reading the Slack core-editor-commits channel.

paaljoachim commented:
Adding in this[ comment](https://github.com/WordPress/gutenberg/issues/29178#issuecomment-860488450) especially:
".....Surely at that point it would be wise to prompt “If you save this block it will update all other uses of it: maybe you want to convert it to an ordinary block?” and offer a Y/N option there. That would have saved me hours of mess."

EDIT: 
Also adding in this [comment](https://github.com/WordPress/gutenberg/issues/32461#issuecomment-860254074):
"Surely if a reusable block has already been used in any other page/post it should be locked and you should be prompted if you try to edit it and warned that any edit will immediately change all other uses of the block? I would expect a dialogue to come up asking if I'm sure I want to edit the reusable block and offering to convert to an ordinary block."

NB! Both of the above comments needs to be addressed in the Locking PR as well as when clicking the Save button.

-----

Keeping the above comment in the back of my mind while I test this PR using this method: https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

After activating the special Gutenberg plugin I received a lot of fatal errors, so that I had to manually remove the plugin from the plugin folder. It would be nice to get this fixed making it easier to test this PR.

shaunandrews commented:
> It would be great to get some additional eyes on this — @pablohoneyhoney, does this asset look ok to you to use in a new Welcome Guide to introduce reusable blocks?

I'm not Pablo, but I like the animation. It could probably be a little smoother; There's some jitter in it now. Some easing could help a little.

critterverse commented:
This is a subtle change but adding some easing definitely helps, thanks for the suggestion @shaunandrews! I also made a minor adjustment to the border radius on the page elements to better match the existing assets. Here's the updated version:

![welcome-reusable](https://user-images.githubusercontent.com/77359364/122603959-34135300-d043-11eb-91a6-c0f07d4a60cc.gif)

@Mamaduka let me know when we have a Trac ticket and I'll add the SVG there, thanks again for your help with the assets 😁

I also had a few quick thoughts about some of the feedback we've been hearing around current confusion re: editing Reusable blocks (some of which was [shared above](https://github.com/WordPress/gutenberg/pull/32464#issuecomment-862172601) by @paaljoachim). I think there are some small adjustments that we could include in this PR that might help:

- Add a more descriptive tooltip on the Save button hover — perhaps it could say something like, "Save globally" or "Update all instances"
- Expand on the description in the Welcome Guide — maybe we could add an additional sentence such as, "To make local changes, try converting to a regular block first."

thisissandip commented:
@Mamaduka I just wanted to ask that if the SVG asset has been uploaded to the SVN repo?

>Add a more descriptive tooltip on the Save button hover — perhaps it could say something like, "Save globally" or "Update all instances"

And, should I go ahead and update the tooltip with "Save globally"?

Mamaduka commented:
Hello, @thisissandip

Sorry, I forgot to create the ticket for this. It's been a busy week.

I will do it first thing on Monday morning.

Mamaduka commented:
We can attach the assets to the old ticket on meta trac and maybe re-open it.

@critterverse, can you attach files since you have all the media assets?

P.S. Sorry for the delay. Feel free to ping me here or on slack if any other action is required from me.

thisissandip committed:
rebase

critterverse commented:
> @critterverse, can you attach files since you have all the media assets?

No problem! I've uploaded the assets and re-opened the original Trac ticket here: https://meta.trac.wordpress.org/ticket/5741

dd32 commented:
Assets uploaded to CDN:

https://s.w.org/images/block-editor/welcome-reusable.gif
https://s.w.org/images/block-editor/welcome-reusable.svg

thisissandip committed:
Add welcome guide image

thisissandip commented:
I have added the Reusable Block Guide Image and I guess now the PR ready. 
Thank you so much for all the help @Mamaduka, @critterverse, @paaljoachim, @dd32, and @jasmussen! 

https://user-images.githubusercontent.com/69596988/124831417-cd75bb00-df98-11eb-869e-b46c75eb9304.mov

paaljoachim commented:
That looks really nice, @thisissandip Sandip!

I shared it in a few of the other Reusable block issues I have. To get any last minute feedback.

pwkip commented:
This still doesn't solve the problem that it's too easy to (accidentally) edit reusable blocks, but at least it's becoming clear that you aren't editing a regular block. I still miss the "Edit" button that used to be there, which you had to click before you could actually start editing the block. Since the save button is in place now, what about changing it to an "Edit" button initially, then once you click it, the block becomes editable and the button changes to the Save button?

I think the cases where you need to edit reusable blocks during the editing process of a post, are rare. I believe reusable blocks need to be edited intentionally, so they should stand out from other blocks in order to not be confused with simple post content.

That said, I'm very happy with this PR and it definitely should be pushed!

paaljoachim commented:
Hey Jules @pwkip 

This is a very interesting idea:

> Since the save button is in place now, what about changing it to an "Edit" button initially, then once you click it, the block becomes editable and the button changes to the Save button?

Sounds like a nice followup PR a little while after this PR has been merged.

critterverse commented:
Thanks so much for all of your work on this so far, @thisissandip! A few notes:

- The Save button height could be a tad bit smaller (32px to make it consistent with the height of other toolbar items)
- Save button tooltip — let’s make this “Save globally” with a lowercase G
- Since we have some extra space in the Welcome Guide, I would still vote for adding the additional sentence mentioned above: "To make local changes, try converting to regular blocks first."

![guide-text](https://user-images.githubusercontent.com/77359364/124962473-a9dc6000-dfec-11eb-8786-7a3f04ca382a.jpg)

- Could you also double check the focus style on the Save button in the toolbar? We seem to be missing the default focus indication there (blue outer ring around the button — please see [this comment](https://github.com/WordPress/gutenberg/pull/32464#issuecomment-861777832) above for reference)

- Lastly, I made a few changes to the asset after seeing the GIF in context — the new version is a bit smoother, crisper, and more closely aligned with the style of the existing animations. (I've annoyingly requested on [Trac](https://meta.trac.wordpress.org/ticket/5741) that the assets be re-uploaded 😅 ).

![welcome-reusable-v2](https://user-images.githubusercontent.com/77359364/124998177-f1c5ac00-e019-11eb-94e5-881f98908e15.gif)

Thanks to everyone who has been contributing to this PR!

thisissandip committed:
welcome guide updates and make save button to 32px in height

critterverse commented:
Adding links to the updated assets here, thanks to @dd32 for uploading!

https://s.w.org/images/block-editor/welcome-reusable.gif?2
https://s.w.org/images/block-editor/welcome-reusable.svg?3

thisissandip commented:
Hi @critterverse, I have made the required changes and updated the assets with the new one.

>Could you also double check the focus style on the Save button in the toolbar? We seem to be missing the default focus indication there (blue outer ring around the button — please see this comment above for reference)

It seems that the outer ring focus is disabled by default for the ToolBar Buttons. Instead, a box inset is used. Please refer to the image below.

![Screenshot 2021-07-12 at 4 03 13 PM](https://user-images.githubusercontent.com/69596988/125313123-90237b80-e352-11eb-8239-e1a5564bcbda.png)

That's why I went ahead and removed the box inset for the Save button in the last commit. I have updated to its default styles again ( the box inset one ).

Please let me know if I should remove the inset or not 😅

critterverse commented:
Interesting, I hadn't noticed before that items in the block toolbar have a slightly different focus style from the default. The most visible place I was able to see the difference is on an active icon — for example when the “bold” option has been pressed in the Paragraph block toolbar:

![focus-style01](https://user-images.githubusercontent.com/77359364/125542098-b6cf9cf0-85be-4714-9122-ca4ecb29f5aa.jpg)

If we apply the inset focus style to the Save button, it looks a little different from similar buttons but matches nicely with other block toolbar items:

![focus-style02](https://user-images.githubusercontent.com/77359364/125541173-cd404a27-6a0f-4407-82b1-241f76a05a12.jpg)

This seems to make the most sense for now (and further unifications to focus styles could potentially be revisited more holistically later if needed). Can we try the inset focus style you've got working already @thisissandip, but matching the width of the button as shown above? The other changes you made look great 👍

thisissandip committed:
Save button box inset

thisissandip commented:
Adjusted the box inset as suggested by @critterverse. 

Now, this is how the whole process of saving a reusable block from the block toolbar looks like: 


https://user-images.githubusercontent.com/69596988/125689083-e04ce2e6-c3a6-48f4-8302-96b4c0376842.mov

thisissandip commented:
Adjusted the box inset as suggested by @critterverse. 

Now, this is how the whole process of saving a reusable block from the block toolbar looks like: 


https://user-images.githubusercontent.com/69596988/125689083-e04ce2e6-c3a6-48f4-8302-96b4c0376842.mov

talldan commented on packages/block-library/src/block/edit.js:
```diff
@@ -85,6 +98,31 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
 	);
 
+	// local states for gudie modal
+	const [ isGudieOpen, setIsGudieOpen ] = useState( false );
```
Small typo here, it should be:
```
const [ isGuideOpen, setIsGuideOpen ] = useState( false );
```

talldan commented on packages/block-library/src/block/edit.js:
```diff
@@ -85,6 +98,31 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+			actions: [
+				{
+					label: __( 'Learn more' ),
+					onClick: () => setIsGudieOpen( true ),
```
This more of a design question (cc @critterverse), but I feel like the snackbar notice is a very brief interaction for it to be the only place to show a guide for reusable blocks.

Saving a reusable block is also towards the end of the process. I wonder if there's another more easily accessed place we'd also consider showing a guide towards the start of the flow of creating a reusable block. The modal where a user names the reusable block was one place I considered, but that would mean having two modals open. The block inspector for the reusable block is another option.

Anyway, this is not something to address in this PR, @thisissandip, so don't worry about it right now. 😄

talldan commented on packages/block-library/src/block/edit.js:
```diff
@@ -127,6 +165,18 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+							isPrimary
+							className="block-library-block__reusable-block-save-button"
+							onClick={ saveEditedRecords }
+							label={ __( 'Save globally' ) }
```
Might need some copy review on this label. Perhaps it could be 'Save reusable block'.

talldan commented on packages/block-library/src/block/editor.scss:
```diff
@@ -14,3 +14,45 @@
+		&::before {
+			left: 0 !important;
+			right: 0 !important;
+			height: 32px !important;
```
There are some grid unit scss variables that should be used in the _variables.scss file (`$grid-unit-40`) for this one and line 22. (edit: I see now they're used elsewhere in the PR, so I guess this was just missed 👍 )

Would be good to add a comment above these lines about why `!important` is needed, what in particular this is overriding.

talldan reviewed (commented):
I haven't tested it, but figured I'd do a code review, looks really good so far. Thanks for working on this, @thisissandip.

critterverse commented:
LGTM, thanks @thisissandip 🎉

>I wonder if there's another more easily accessed place we'd also consider showing a guide towards the start of the flow of creating a reusable block.

I like this suggestion @talldan — opened a follow up issue in #33481. Thanks for the code review!

thisissandip committed:
fix typos and add css comments

thisissandip commented:
Thank you for the code review @talldan! 😄 

I have fixed the typos and made the required changes to the scss file. Let me know if this looks good.

kellychoffman commented:
Looking good. The biggest issue is that the Save option is completely hidden and there's no way to know to click on that block icon to access the Save button. As a stop gap - What if we add a dot just like we add a dot to the update button in the top toolbar? I think that would at least guide the user to click on it. Something like:

<img width="542" alt="Screen Shot 2021-07-27 at 7 31 41 AM" src="https://user-images.githubusercontent.com/5835847/127172753-a9fdf976-99f4-46e6-921b-3993bba7901e.png">

thisissandip commented:
Hi @kellychoffman, please correct me if I am wrong 😅 

So if there are unsaved changes, a dot will prepend the parent block selector icon, right?

Something like this:

![image](https://user-images.githubusercontent.com/69596988/127102506-68275e49-efde-4282-8f2a-3aa42e458ecb.png)

critterverse commented:
Tried out a few different ways that we could indicate unsaved changes in the parent selector without having to change its size/footprint:

![save-dot](https://user-images.githubusercontent.com/77359364/127195664-c6a3cd37-69bf-40a3-80a4-bfd0e0c3cce7.jpg)

But I think Kelly's [proposal above](https://github.com/WordPress/gutenberg/pull/32464#issuecomment-886911336) works best/is the most consistent with the way dots appear in other locations!

shaunandrews commented:
> without having to change its size/footprint:

FWIW, I like that the parent button gets a little larger – it helps suggest that something is different.

talldan commented:
@kellychoffman @critterverse @shaunandrews Would be good to get a steer on whether the indicator is critical for this PR, or whether it could be tackled in a separate change?

This might be a deceptively complex thing to implement—the parent block selector button is just a generic button, it might be challenging from a technical standpoint to make it aware of whether the parent block has its own save mechanism as conceptually that's encapsulated within the reusable block.

thisissandip commented:
@talldan I gave this a shot and was able to implement this. 

Here's how it works:

https://user-images.githubusercontent.com/69596988/127275144-f16715fc-b687-461a-95c1-871e2c0a6206.mov

Should I push the changes to this PR or wait for a new issue to be created after this one is merged?

talldan commented:
@thisissandip I think you should be able to create a separate PR that merges into this branch, that might be a good way to do things.

thisissandip committed:
add dot to parent selector for reusable block

thisissandip commented:
@talldan I have added a PR https://github.com/thisissandip/gutenberg/pull/1

kellychoffman commented:
@thisissandip Nice work! 

@talldan Given the name + scope of this PR, I think it's worth including here as it is part of the flow and without its very unclear that you have to click on it to access the save button. Either this or we rethink the whole flow and I don't think that is a practical path right now - better to get this out as a full flow to see how it performs.

critterverse commented:
One small suggestion for the dot flow — could we change it so that the dot remains visible in the Reusable block toolbar after the parent selector is clicked? I think this could help reinforce what the dot is supposed to mean by seeing it in the same view as the Save button. I find it especially helpful to see it disappear from the toolbar and main Publish/Update button at the same time after clicking Save:

https://user-images.githubusercontent.com/77359364/127864889-c4f6eb8c-3cda-4045-aabe-da68c7cf8e1b.mov

cc @kellychoffman

kellychoffman commented:
> the dot remains visible in the Reusable block toolbar after the parent selector is clicked? 

Yes! I think that makes the most sense.

thisissandip committed:
Add dot to block parent seletor and reusable block toolbar

thisissandip commented:
@talldan I have implemented the dot feature without using the core-data library in the block editor package as it was suggested here https://github.com/thisissandip/gutenberg/pull/1#discussion_r678804688

I have updated this PR https://github.com/thisissandip/gutenberg/pull/1 

I'd really appreciate it if I can get a code review for https://github.com/thisissandip/gutenberg/pull/1, so I can go ahead and merge https://github.com/thisissandip/gutenberg/pull/1 into this PR branch 🙇 

Here's how it works:

https://user-images.githubusercontent.com/69596988/128136519-4c5d5ca0-ec91-404a-a9d1-ae78762c33b1.mov

@kellychoffman @critterverse Let me know if the flow looks good or if there are any changes.

critterverse commented:
Looks great, thanks @thisissandip!

critterverse commented:
Hey all, I expanded on the dot indicator logic a little more over in #33988 — nothing that should hold up this PR but it could be a nice follow up! I'd love to be able to apply the learnings from this Reusable block saving flow to other types of blocks 😁

thisissandip committed:
add two separate components for unsaved changes indicator

thisissandip committed:
undo block switcher changes

thisissandip commented:
Hi @talldan,

I have made two separate components for the unsaved changes indicator as suggested by you:
1. ParentBlockSelectorUnsavedChangesIndicator
2. SelectedBlockUnsavedChangesIndicator

These Indicator components can be extended for Template Parts and several other blocks as well as proposed in https://github.com/WordPress/gutenberg/issues/33988 

Let me know if https://github.com/thisissandip/gutenberg/pull/1 looks good. So I can go ahead and merge https://github.com/thisissandip/gutenberg/pull/1 with this PR branch.

Thank you for reviewing and guiding me 😄

kellychoffman commented:
This feels good to ship from a UX perspective! @talldan Anything blocking this from a code perspective?

talldan commented:
I've left a couple more comments on https://github.com/thisissandip/gutenberg/pull/1, but looks like it's technically very close.

Worth flagging this with other members of @WordPress/gutenberg-core, as it introduces a new experimental API for blocks.

thisissandip commented:
@talldan, I have added a [comment](https://github.com/thisissandip/gutenberg/pull/1/files#r696051897) highlighting why the other parent elements (contextual toolbar) needs to be adjusted when the unsaved changes indicator is present inside the parent selector button.

mtias commented:
This is looking good, thank you everyone for all the tweaks and feedback.

Can we consider not rendering the save button at all when there are no changes?

critterverse commented:
>Can we consider not rendering the save button at all when there are no changes?

I think this could work nicely:

https://user-images.githubusercontent.com/77359364/131169412-81a8c1c2-81a7-4609-b20c-7b1c1f3da3ae.mov

@thisissandip if you have a chance, it might be worth incorporating this logic first and then we can see if there needs to be some animation added to make it transition more smoothly. One thing that comes to mind is that when saving, the dot should remain visible during the button's busy state — that way both could quickly fade out + the toolbar collapse down all at the same time once the action is complete.

paaljoachim commented:
Hello everyone. This is after a few months perspective of seeing various feedback in relation of the usage of Reusable blocks.

Checking out @critterverse Channing's above video I noticed the dot next to the Reusable block icon. Including the dot there is a very helpful hint to show that changes has happened. Issue: https://github.com/WordPress/gutenberg/issues/33988

### Undo not working properly.
I made a change and clicked ctrl+z (I could have clicked the top left document undo icon) and noticed the change I made undone. BUT the publish button still has a dot in it. The dot should be removed and the option to save Reusable block should not be present, because I undid the change. Meaning there is nothing to save.
--> Needs a followup in another PR.

I downloaded the [Gutenberg PR build ](https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/)to do a status check to see where the PR is at currently. (Which means commenting on this that have likely also already been mentioned.)

Clicking the overlay. Selects the Reusable block.
I see a Save button. As Matias mentioned: 

> Can we consider not rendering the save button at all when there are no changes?

It is a good idea to remove the save button when there are no changes to save.


### Change the text in the tooltip.
I hovered over the Save button and saw this:
![Screen Shot 2021-08-28 at 00 07 04](https://user-images.githubusercontent.com/5323259/131193423-a222fc3a-bc0f-4c32-bc8b-3aadb4a0e963.jpg)

The tooltip says: "Save reusable block." 
Should it instead say? "Save changes to Reusable block." OR "Save changes to everywhere this Reusable block is used."
(Yep might be too long). As it gives a hint that clicking save has a much larger impact than the user might know about.


### Learn more (Wondering about the usefulness of having it show up after a save.)
Clicking Save. One sees this:
![Screen Shot 2021-08-28 at 00 08 56](https://user-images.githubusercontent.com/5323259/131193570-22d05e4d-bdce-4469-a291-ae9790be9de4.jpg)

I know that @critterverse Channing has mentioned changing the Reusable block description seen in the block settings sidebar and adding a link to documentation. That would be more helpful. (At the moment I am not finding the issue for it.)

Because having a Learn more link after having saved a change to a Reusable block is a bit late in learning what a Reusable block does. If one does not know what one is doing at this stage one might have just destroyed a Reusable block used across the site. I am not certain how useful a Learn more link is at this stage. I would be better to have a preventative method by adding text to the description of the block as Channing is working on.

An Option! Having the notification with a Learn more show up right after converting to a Reusable block.

That means changing this notification which shows up after clicking Add to Reusable blocks and giving it a name and clicking Save. Going from:
![Converting-to-Reusable-block-notification](https://user-images.githubusercontent.com/5323259/131198287-86dc99c7-212e-4845-8274-d61d07426535.jpg)

To become **"Reusable block created. Learn more"**
![Reusable-block-created-Learn-more-notification](https://user-images.githubusercontent.com/5323259/131198599-a94243b8-db73-4272-8182-f63efab11de6.jpg)

As the user will be able to learn more about Reusable blocks after just having created one.



### Save buttons
A kind of quick Save button in the toolbar. (Can be risky for people who do not know what they are doing. As it makes it very easy to quickly save a change to a type of block that has a much wider impact than just the page or post on is in.)
A Save button in the Publish panel. (Can be used to add information about the impact of saving a change to a Reusable block across the site to everywhere it might be used.)

thisissandip committed:
hide save btn when there are no changes

thisissandip commented:
Apologies for the late reply. I was AFK for the week

>Can we consider not rendering the save button at all when there are no changes?

@mtias Yes, this makes the most sense.

I have updated the PR branch https://github.com/thisissandip/gutenberg/pull/1 with the recent explorations

Now the process of saving a reusable block looks like this:

https://user-images.githubusercontent.com/69596988/131361333-8ff11db5-772d-44d2-bce8-196984ef4937.mov

> @thisissandip if you have a chance, it might be worth incorporating this logic first and then we can see if there needs to be some animation added to make it transition more smoothly.

@critterverse Adding animations might be worth exploring cause the current process feels kind of "_snappy_"


**UPDATE**: I don't know why the Tests are failing for https://github.com/thisissandip/gutenberg/pull/1 even though all the tests passed locally.

critterverse commented:
Thanks @thisissandip! Here's a quick motion test of how the animation could work (the dot and Save button fade out first and then the toolbar collapes down):

![toolbar-collapse](https://user-images.githubusercontent.com/77359364/131363155-0b9db196-c05f-4d09-a7d4-83437382f9fd.gif)

Also thanks for the in depth review, @paaljoachim!

>Because having a Learn more link after having saved a change to a Reusable block is a bit late in learning what a Reusable block does. An Option! Having the notification with a Learn more show up right after converting to a Reusable block.

This is a good point that did come up once before. Let's discuss over in https://github.com/WordPress/gutenberg/issues/33481 because I think there are a few different follow-up options that would make this messaging available earlier in the process and compliment what's in this PR.

desrosj commented:
Just noting that this one will likely miss WordPress 5.8.1. RC is due out at some point on Wednesday. And I think I'd like to see this receive some time for testing in the plugin before considering including it in a minor release of WordPress.

I'll leave the label so it is not overlooked in consideration for 5.8.2. Please also feel free to remove the `Backport to WP Minor Release` label if this has turned into something that should wait for WP 5.9.

paaljoachim commented:
Thank you @desrosj Jonathan!

I removed the backport to minor label, as there are various discussions spun out from this PR that we also need to consider.

thisissandip committed:
Merge pull request #1 from thisissandip/add/dot-to-parent-selector

Add a dot before the parent block selector when there are unsaved changes in the Reusable block

thisissandip commented:
Merged https://github.com/thisissandip/gutenberg/pull/1 to this PR. 

It adds the unsaved changes indicator ( a dot )  to the Parent Selector Button of the Inner blocks and the Block Toolbar of the Reusable block.

**UPDATE**: It is mentioned in the [core-editor channel](https://wordpress.slack.com/archives/C02QB2JS7/p1630401838127800) that the failing tests are because of https://github.com/WordPress/gutenberg/issues/34377

thisissandip commented:
@critterverse That looks really good!

Implementing an animation like that in the block toolbar without an animation library is quite complex. 

I'll appreciate any help/suggestions to achieve this 😄

kellychoffman commented:
> Implementing an animation like that in the block toolbar without an animation library is quite complex.

I think we can move forward on this PR without the animation. We can address that separately.

mtias commented:
Yes, should not be blocking the initial rollout, I can take a look for the animation implementation later on.

thisissandip committed:
update label and adjust margin

thisissandip commented:
I have updated the label to "Save Globally" and increased the margin of the save button since the button was not centered correctly within the Toolbar group as suggested by @critterverse

jasmussen commented on packages/block-library/src/block/editor.scss:
```diff
@@ -18,7 +18,7 @@
 .components-toolbar-group {
 	.block-library-block__reusable-block-save-button {
-		margin: 7px;
+		margin: 8px;
```
You can use $grid-unit-10 instead of 8px if you like.

thisissandip replied:
Oh, Yes! Thank you! 😄

thisissandip committed:
use grid-unit-10

critterverse commented:
Tested in Chrome, Firefox, and Safari and all is working well ✨ 

Small note that on the parent selector, it looks like the focus indicator is cutting off the dot on the left edge a tiny bit. It looks ok on the block toolbar though where the spacing should be the same:

<img width="710" alt="RB-parent-selector" src="https://user-images.githubusercontent.com/77359364/132587785-3cd0cad1-d811-4fd7-9e40-1c48d8e8fbb2.png">
<img width="710" alt="RB-toolbar" src="https://user-images.githubusercontent.com/77359364/132587787-364c3bed-2a42-4a07-b76f-fceb3cf6ab1f.png">

thisissandip committed:
adjust unsaved changes indicator margin

thisissandip commented:
Thank you for pointing this out @critterverse!

I've updated the margin of the unsaved changes indicator in the parent selector button.  Now it looks like this when focused:

![image](https://user-images.githubusercontent.com/69596988/132704948-61f64683-e84c-46fc-af11-f1b9d6d71651.png)

critterverse commented:
>Small note that on the parent selector, it looks like the focus indicator is cutting off the dot on the left edge a tiny bit.

@thisissandip Sorry for not being very clear in my feedback above! To clarify, I think we could make an adjustment to the focus indicator versus adjusting the dot placement. Otherwise it causes the dot to jump around when moving between the toolbar and parent selector:

![dot-indicator](https://user-images.githubusercontent.com/77359364/133132595-c62946ee-f3bd-4b8f-977b-1c2ad0e90d93.gif)

Here's the spacing specs for easy reference:

**EDIT: Revised the specs below to match the current spacing used in this PR and deleted a comment referencing outdated spacing**

<img width="100%" alt="toolbar-specs" src="https://user-images.githubusercontent.com/77359364/135880704-13801ac9-d78a-472c-89ef-ad77d19efb1e.png">

Re: Parent selector focus indicator — I wonder if we could resolve the issue with the dot being cut off by making the focusable area 2px wider? 🤔

Before adjustment:

![no-change](https://user-images.githubusercontent.com/77359364/133160232-a3df0dd3-7ffd-4f69-b906-575feeabf784.png)

After adjustment:

![updated](https://user-images.githubusercontent.com/77359364/133159766-4e1f59b5-1b40-4c06-8649-571de7cc2d3d.png)

thisissandip committed:
increase parent selector focus indicator width

thisissandip commented:
Hi @critterverse, thanks for the clarification!

> Re: Parent selector focus indicator — I wonder if we could resolve the issue with the dot being cut off by making the focusable area 2px wider? 🤔

Increasing the Parent selector focus indicator area by 1px solved the issue.

![Screen-Recording-2021-09-14-at-8](https://user-images.githubusercontent.com/69596988/133285597-5216c5ab-9df9-45e5-b302-e5a2ceec2d4b.gif)

critterverse commented:
>Increasing the Parent selector focus indicator area by 1px solved the issue.

Awesome, thanks @thisissandip! 

Looks like the icon is jumping around a bit when switching between the parent selector and toolbar. A couple of suggestions:

**Parent selector:**
Change the `width` of the parent selector button from 64px to 62px. 

**Toolbar:**
Change the `margin-right` of the dot indicator from 8px to 6px.

I think that should fix the alignment :)

ntsekouras commented on packages/block-library/src/block/reusable-block-welcome-guide.js:
```diff
@@ -0,0 +1,56 @@
+import { Guide } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+function WelcomeGuideImage( { nonAnimatedSrc, animatedSrc } ) {
```
I have mixed feelings about this as well since we are not introducing a new concept here and `save` button/action used to be there before.. Also I've tried your PR a bit yesterday and today and both times I loaded a post the first time, it showed me the basic welcome guide. I didn't investigate why this happened though. Did this happen for anyone else?

thisissandip replied:
The welcome guide will show up only after clicking the "Learn more" option in the snackbar like this:

![Reusable block welcome guide](https://user-images.githubusercontent.com/69596988/134027636-360480fc-3a12-4a63-a6f6-f042f6cb657f.gif)

ntsekouras commented on packages/block-library/src/block/edit.js:
```diff
@@ -47,14 +57,37 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+			isChildSelected: _isChildSelected,
+			isSelected: _isSelected,
+		};
+	}, [] );
```
`clientId` is a dependency here.

thisissandip replied:
Thank you! You saved me from future errors! 🚀

ntsekouras commented on packages/block-library/src/block/edit.js:
```diff
@@ -47,14 +57,37 @@ export default function ReusableBlockEdit( { attributes: { ref }, clientId } ) {
+
+		const parents = getBlockParents( selectedBlockClientId );
+		const firstParentClientId = parents[ parents.length - 1 ];
+		const _isChildSelected = firstParentClientId === clientId;
```
Shouldn't the case of changing nested blocks inside the `reusable` block been handled somehow? I was expecting here to see a `getBlockRootClientId` and check if is `reusable` and `hasEdits`... If we add a `Group` with a `Paragraph` inside the reusable block and change that paragraph, the block has edits but there is no indication for the changes. Of course in this case the `Group` block is the parent.. I'm not sure how it would be best to show this.

thisissandip replied:
> I was expecting here to see a getBlockRootClientId and check if is reusable and hasEdits

Yes, this is much cleaner and better. Thank you for the suggestion. Updating this in my next commit.

> If we add a Group with a Paragraph inside the reusable block and change that paragraph, the block has edits but there is no indication for the changes. Of course in this case the Group block is the parent.. I'm not sure how it would be best to show this.

Even I am not sure about this case cause one has to select the Reusable block itself to access the save button in the block toolbar.

critterverse replied:
>If we add a Group with a Paragraph inside the reusable block and change that paragraph, the block has edits but there is no indication for the changes. 

Introducing dot indicators into List view has come up as one possible solution that would help with more visibility in instances with a lot of nesting. Needs a design explore but I could look at this alongside other follow ups in https://github.com/WordPress/gutenberg/issues/33988!

ntsekouras reviewed (commented):
Thanks for your work here @thisissandip and for the iterations! Cool work! 💯 

When I first saw the problem that you're trying to solve I thought that a custom hook would be good and then we could use that for other similar blocks, like `Template Parts` maybe.. Unfortunately this is not so easy as we need a dependency of `core-data` for `hasEditsForEntityRecord`, unless there is another way I didn't think of.. 🤔 

Regarding the `slots` I'm a bit concerned creating two of them and so specific to the thing they add (`unsaved changes indicator`). Do we expect something else to be added before the `block icon`? Should we rename them to something more generic?

I don't have a clear proposal for these - just some thoughts of mine 😄 

Also have you tried using only one slot? Could this work? And lastly without seeing all the css code, could some css added with `:after` (or something similar) without adding a new `div` node, could work and simplify things?

thisissandip commented:
> Looks like the icon is jumping around a bit when switching between the parent selector and toolbar.

@critterverse While I was investigating this, I found that the jump of the icon occurs on different screen widths. For example:

Screen width 942px — icon jumps
Screen width 943px — no jump
Screen width 944px — icon jumps
Screen width 945px — no jump

and again

Screen width 1238px — icon jumps
Screen width 1237px — no jump

It seems the jump occurs only when the width of the screen size is an even number. 

This is a weird issue, I can't figure out the reason. (Maybe because of my monitor's low resolution?)

Can you confirm if you can replicate this on your side?

thisissandip commented:
@ntsekouras Thank you for taking a look and for the code review! 😄 


> Regarding the slots I'm a bit concerned creating two of them and so specific to the thing they add (unsaved changes indicator). Do we expect something else to be added before the block icon? Should we rename them to something more generic?

> Also have you tried using only one slot? Could this work?

I gave this a shot and was able to implement this using only one slot and a single Unsaved Changed Indicator component called `BlockUnsavedChangesIndicator`. 

I have created a separate PR: https://github.com/thisissandip/gutenberg/pull/2 in my forked repository to keep things clean.

I will really appreciate it if you can take a look so maybe we can drop using 2 separate slots and 2 different components for the unsaved changed indicator.


> And lastly without seeing all the css code, could some css added with :after (or something similar) without adding a new div node, could work and simplify things?

Adding `:after` or `:before` to the block icon decreases the block icon size and creates a situation like this:

![image](https://user-images.githubusercontent.com/69596988/134070109-424c329b-d4e9-4e39-aacc-1bd38f3fdc1d.png)


That's why I added a div for the dot (unsaved changes indicator)

critterverse commented:
>It seems the jump occurs only when the width of the screen size is an even number. This is a weird issue, I can't figure out the reason. 

Oh weird! Yes, I can replicate...

>Maybe because of my monitor's low resolution?

FWIW I'm seeing the jump even when testing on a retina screen 🤔

critterverse commented on packages/block-editor/src/components/block-parent-selector/style.scss:
```diff
@@ -9,3 +9,36 @@
+
+	// move the parent selector icon to right side to make room for the dot
+	.block-editor-block-icon {
+		margin-right: -15px;
```
Changing this value to -16px (or `$grid-unit-20`) seems to get rid of the visual jump for me (tested on a couple of different screens at different viewport sizes).

kellychoffman committed:
gets rid of visual jump

Props @critterverse 
Ref: https://github.com/WordPress/gutenberg/pull/32464#pullrequestreview-770494293

kellychoffman commented:
This has sat for a bit but looks good from a UX perspective. Can we land this?

paaljoachim commented:
Let's land this outside of WordPress 5.9, as it feels like one part of a bigger change.
(Meaning let's land this at earliest next week. After the feature freeze date for 5.9 in relation to the Gutenberg plugin is on Friday the 5th of November.)

talldan commented:
@thisissandip Is this something you're still interested in progressing? I realise the scope of this change increased dramatically, but it is a really valuable change, so if you're still interested that would be really great.

If you don't have time or you have other commitments, would you be ok with another community member working on this? Ideally it'd be done in a way where you still get props for the majority of the work.

talldan commented on packages/block-editor/src/components/block-unsaved-changes-indicator/parent-block-selector-unsaved-changes-indicator.js:
```diff
@@ -0,0 +1,42 @@
+
+function UnsavedChangesIndicator() {
+	useEffect( () => {
+		// add class to the parent selector to increase it's width
+		const parentSelectorBtn = document.querySelector(
+			'.block-editor-block-parent-selector'
+		);
+		parentSelectorBtn.classList.add( 'parent-block-has-changes' );
+		// add class to the contextual toolbar to move it to the right side
+		const contextualToolBar = document.querySelector(
+			'.block-editor-block-contextual-toolbar'
+		);
+		contextualToolBar.classList.add( 'parent-block-has-changes' );
+
+		return () => {
+			// remove classes from the parent selector and contextual toolbar when component unmounts
+			parentSelectorBtn.classList.remove( 'parent-block-has-changes' );
+			contextualToolBar.classList.remove( 'parent-block-has-changes' );
+		};
```
Something that I mentioned on the other PR (https://github.com/thisissandip/gutenberg/pull/1), I think it'd be good to make this slot not modify the element's class names in this way. This couples the `UnsavedChangesIndicator` to the class name of the parent selector button, if that classname changes then this breaks.

I recall it was because the parent selector button is positioned absolutely. That seems unusual to me, and like some tech debt. Maybe those toolbar styles can be refactored.

thisissandip commented:
Hi @talldan 👋 ,

> If you don't have time or you have other commitments, would you be ok with another community member working on this? Ideally, it'd be done in a way where you still get props for the majority of the work.

I am so sorry for not being able to update this PR as I am busy with some other commitments. It would be okay with me if someone else pick this up from here 🙂 

I just want to make a note of the things that are left for this PR:
1. The [dot indicator is jumping around a bit](https://github.com/WordPress/gutenberg/pull/32464#issuecomment-918629320) when switching between the parent selector and toolbar. This can be solved by making minor adjustments in CSS. As suggested by @critterverse [here](https://github.com/WordPress/gutenberg/pull/32464#discussion_r721507936).
2. The `UnsavedChangesIndicator` should not modify the class name of the parent selector button [here](https://github.com/WordPress/gutenberg/pull/32464/files/740a109caf0bdbb4bc912b5d800f191f46d9a99c#r741672902) because it couples the `UnsavedChangesIndicator` with the class name of the parent selector button which can lead to breaking changes as pointed out by @talldan.

Once again, I am sorry for not being able to work on this PR at the moment. I'll come back to contribute to Gutenberg again in a few weeks!

critterverse commented:
>The dot indicator is jumping around a bit when switching between the parent selector and toolbar. This can be solved by making minor adjustments in CSS. 

@kellychoffman added this fix in https://github.com/WordPress/gutenberg/pull/32464/commits/740a109caf0bdbb4bc912b5d800f191f46d9a99c so I think the only remaining issue is the one @talldan [mentioned](https://github.com/WordPress/gutenberg/pull/32464#discussion_r741672902).

>I am so sorry for not being able to update this PR as I am busy with some other commitments.

Totally understandable, @thisissandip. Thank you so much for all the great work on this PR and for seeing it through across so many iterations! 👏

talldan commented:
The related issue has now been closed, so I'll close the PR as well. Thanks for all the explorations here.
